### PR TITLE
Support LogQL HTTP query APIs.

### PR DESCRIPTION
**Pull request by weixiang1862**
### LogQL Service
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [x] Update the documentation to include this new feature.
- [x] Tests(including UT, IT, E2E) are added to verify the new feature.
- [x] If it's UI related, attach the screenshots below.
d0d453718453e01fa1d6ab0f1a50bed

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #11130.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

**Comment by weixiang1862** [comment]
Some pictures refer in docs now contains in this commit, once the docs reviewed, I will open another pr in `skywalking-website`and migrate them to there.

**Comment by wu-sheng** [comment]
Please post the picture by commenting, then we know what they are.

File: docs/en/setup/backend/ui-grafana.md [hunk 1]
```diff
@@ -55,6 +60,27 @@ General settings:
...
+
+### Add Log Panel
+The following steps show how to add a log panel.
+1. Choose `Logs` chart.
+2. Set the `Line limit` value (The max number of logs to return in a query) and `Order` value (Determines the sort order of logs).
+3. Add LogQL expressions, use the variables configured above for the labels and searching keyword.
+4. Test query and save the panel.
+<img src="./grafana-log-panel.jpg"/>
+
 ## Preview on demo.skywalking.a.o
```
Comment (wu-sheng) [inline docs/en/setup/backend/ui-grafana.md:84]: Preview section is also expected later.

**Comment by weixiang1862** [comment]
> Please post the picture by commenting, then we know what they are.

They are pictures in [docs/en/setup/backend/ui-grafana.md](https://github.com/apache/skywalking/pull/11168/files/e1d8fcb9e3a825197e5994c33e23d7b1429d1d0c#diff-46c8605fc482404cbe57a479d7b3f11ccb38d691b1f1100cb8d485fdd333f173).

grafana-loki-datasource

grafana-loki-variables1

grafana-loki-variables2

grafana-loki-variables3

grafana-log-panel

grafana-loki-variables4

File: oap-server/server-starter/src/main/resources/application.yml [hunk 2]
```diff
@@ -431,6 +431,18 @@ promql:
...
     restIdleTimeOut: ${SW_PROMQL_REST_IDLE_TIMEOUT:30000}
     restAcceptQueueSize: ${SW_PROMQL_REST_QUEUE_SIZE:0}
+    
+#This module is for LogQL API.
+logql:
+  selector: ${SW_LOGQL:default}
+  default:
+    # For HTTP server
+    restHost: ${SW_LOGQL_REST_HOST:0.0.0.0}
+    restPort: ${SW_LOGQL_REST_PORT:3100}
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/application.yml:441]: With this, `showcase` and `helm` codes need to be updated to expose the new port, right? @kezhenxu94 @wankai123

File: oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/handler/LogQLApiHandler.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        LogQLExprVisitor visitor = new LogQLExprVisitor();
+        LogQLParseResult parseResult = visitor.visit(tree);
+        Map<String, String> labelMap = parseResult.getLabelMap();
+
+        String serviceId = labelMap.containsKey(LabelName.SERVICE.getLabel()) ?
+            IDManager.ServiceID.buildId(labelMap.get(LabelName.SERVICE.getLabel()), true) : null;
+        String serviceInstanceId = labelMap.containsKey(LabelName.SERVICE_INSTANCE.getLabel()) ?
+            IDManager.ServiceInstanceID.buildId(serviceId, labelMap.get(LabelName.SERVICE_INSTANCE.getLabel())) : null;
+        String endpointId = labelMap.containsKey(LabelName.ENDPOINT.getLabel()) ?
+            IDManager.EndpointID.buildId(serviceId, labelMap.get(LabelName.ENDPOINT.getLabel())) : null;
```
Comment (wankai123) [inline oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/handler/LogQLApiHandler.java:148]: If `serviceId` is null, build instanceId and endpoint id will get unexpected id `_blank`.

**Review by wankai123 — CHANGES_REQUESTED** [review]

File: oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/handler/LogQLApiHandler.java (hunk 3 above)
Comment (weixiang1862) [reply oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/handler/LogQLApiHandler.java:148]: Ok, thanks. I will add a check.

File: oap-server/server-starter/src/main/resources/application.yml (hunk 2 above)
Comment (kezhenxu94) [reply oap-server/server-starter/src/main/resources/application.yml:441]: I suggest we don't enable logql, and promql which I missed before, they are occupying too many ports and are not common cases where users would use, just let the users enable these when they want

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/application.yml:441]: I am fine about enabling or not by default. It seems not much different. Are you concern about network security impact?

File: test/e2e-v2/cases/logql/logql-cases.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  - query: curl -X GET http://${oap_host}:${oap_3100}/loki/api/v1/labels -d 'start='$(($(($(date +%s)-1800))*1000000000))'&end='$(($(date +%s)*1000000000))
+    expected: expected/logql-log-tags.yml
+  # log tag value query
+  - query: curl -X GET http://${oap_host}:${oap_3100}/loki/api/v1/label/level/values -d 'start='$(($(($(date +%s)-1800))*1000000000))'&end='$(($(date +%s)*1000000000))
+    expected: expected/logql-log-tag-values.yml
+  # log range query FORWARD
+  - query: curl -X GET http://${oap_host}:${oap_3100}/loki/api/v1/query_range -d 'query={service="e2e-service-provider"}&start='$(($(($(date +%s)-1800))*1000000000))'&end='$(($(date +%s)*1000000000))'&limit=100&direction=FORWARD'
+    expected: expected/logql-logs.yml
+  # log range query BACKWARD
+  - query: curl -X GET http://${oap_host}:${oap_3100}/loki/api/v1/query_range -d 'query={service="e2e-service-provider"}&start='$(($(($(date +%s)-1800))*1000000000))'&end='$(($(date +%s)*1000000000))'&limit=100&direction=BACKWARD'
```
[the remark this answers is not in the document]
Comment (weixiang1862) [reply test/e2e-v2/cases/logql/logql-cases.yaml:27]: Ok, I will add more cases.

File: oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/entity/LabelName.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+package org.apache.skywalking.oap.query.logql.entity;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+@JsonSerialize()
```
[the remark this answers is not in the document]
Comment (weixiang1862) [reply oap-server/server-query-plugin/logql-plugin/src/main/java/org/apache/skywalking/oap/query/logql/entity/LabelName.java:27]: Yes, you all right. It is useless in logql.

File: test/e2e-v2/cases/logql/logql-cases.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (weixiang1862) [reply test/e2e-v2/cases/logql/logql-cases.yaml:27]: Added.

File: docs/en/setup/backend/ui-grafana.md [hunk 6]
```diff
@@ -55,6 +60,27 @@ General settings:
...
 
+## Configure Log Dashboard
+### Dashboards Settings
+The following steps are the example of config a log dashboard:
+1. Create a dashboard named `Log`.
+2. Configure variables for the dashboard:
+<img src="./grafana-loki-variables1.jpg"/>
+3. Please make sure `service_instance` and `endpoint` variable enabled `Include All` option and set `Custom all value` to blank (typed by space button on the keyboard):
+<img src="./grafana-loki-variables2.jpg"/>
+4. `Tags` variable is a little different from others, for more details, please refer [Ad hoc filters](https://grafana.com/docs/grafana/latest/dashboards/variables/add-template-variables/#add-ad-hoc-filters):
```
Comment (wu-sheng) [inline docs/en/setup/backend/ui-grafana.md:71]: `Ad hoc filters` markdown format seems not working well. 

image

It should be a link. Maybe the `:` breaks it? Please recheck.
The preview is on GitHub preview page. https://github.com/apache/skywalking/blob/e519edc71c8635f8a3dde9fdffc7af8e8b3da6d2/docs/en/setup/backend/ui-grafana.md

File: oap-server/server-starter/src/main/resources/application.yml (hunk 2 above)
Comment (kezhenxu94) [reply oap-server/server-starter/src/main/resources/application.yml:441]: > I am fine about enabling or not by default. It seems not much different. Are you concern about network security impact?

it's much different, enabling an unused plugin creates an extra HTTP server, consuming more resources, and security impact is the most important concern of mine.

**Comment by wu-sheng** [comment]
I feel good about these screenshots on the Grafana, @wankai123 WDTY? If you agree, I would ask these to be hosted on website repo.

**Comment by wankai123** [comment]
> I feel good about these screenshots on the Grafana, @wankai123 WDTY? If you agree, I would ask these to be hosted on website repo.

image
If the ServiceInstance or Endpoint is `All`, will serviceInstance and endpoint label pass to the API handler？ Then backend will use `All` as the name to build a not existing Id.

**Comment by kezhenxu94** [comment]
> > I feel good about these screenshots on the Grafana, @wankai123 WDTY? If you agree, I would ask these to be hosted on website repo.
> 
> 
> 
> image
> 
> If the ServiceInstance or Endpoint is `All`, will serviceInstance and endpoint label pass to the API handler？ Then backend will use `All` as the name to build a not existing Id.

I think `All` in the Grafana UI usually means this field has no query condition, instead of sending an `All` as a query condition.

**Comment by wankai123** [comment]
> I think `All` in the Grafana UI usually means this field has no query condition, instead of sending an `All` as a query condition.

If so, I'm good.

**Comment by weixiang1862** [comment]
And grafana doesn't support empty value variable, here is the discussion:
https://github.com/grafana/grafana/issues/5970

**Comment by wankai123** [comment]
> The all option in grafana submit all options sperate by `,` and it is not we wanted. So I custom the all value to `blank` and backend filter the blank label value(service, endpoint): image

`typed by space button on the keyboard` seems a little weird, WDYT @wu-sheng , do we need to define a specific name for `All`?

**Comment by kezhenxu94** [comment]
> > > > I feel good about these screenshots on the Grafana, @wankai123 WDTY? If you agree, I would ask these to be hosted on website repo.
> 
> > > 
> 
> > > 
> 
> > > image
> 
> > > If the ServiceInstance or Endpoint is `All`, will serviceInstance and endpoint label pass to the API handler？ Then backend will use `All` as the name to build a not existing Id.
> 
> > 
> 
> > I think `All` in the Grafana UI usually means this field has no query condition, instead of sending an `All` as a query condition.
> 
> 
> 
> The all option in grafana submit all options sperate by `,` and it is not we wanted. So I custom the all value to `blank` and backend filter the blank label value(service, endpoint):
> 
> image
> 

Can we customize the all value to `*`? That's much more understandable

File: docs/en/setup/backend/ui-grafana.md (hunk 6 above)
Comment (weixiang1862) [reply docs/en/setup/backend/ui-grafana.md:71]: I adjust it in many ways, it fails. Seems github preview page cant not rendering a link in markdown list. It rendering well in idea and other markdown editor.

**Comment by weixiang1862** [comment]
> > > > > I feel good about these screenshots on the Grafana, @wankai123 WDTY? If you agree, I would ask these to be hosted on website repo.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > image
> > 
> > 
> > > > If the ServiceInstance or Endpoint is `All`, will serviceInstance and endpoint label pass to the API handler？ Then backend will use `All` as the name to build a not existing Id.
> > 
> > 
> > > 
> > 
> > 
> > > I think `All` in the Grafana UI usually means this field has no query condition, instead of sending an `All` as a query condition.
> > 
> > 
> > The all option in grafana submit all options sperate by `,` and it is not we wanted. So I custom the all value to `blank` and backend filter the blank label value(service, endpoint):
> > image
> 
> Can we customize the all value to `*`? That's much more understandable

Value `*` or other special character is ambiguity, `{service_instance="*"}` or `{service_instance="_all_"}` in logql means it want filter a log with label value `*` or `_all_`,.
I think blank value is reasonable than others in semantics, although it is weird.

**Comment by weixiang1862** [comment]
> > > > > I feel good about these screenshots on the Grafana, @wankai123 WDTY? If you agree, I would ask these to be hosted on website repo.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > image
> > 
> > 
> > > > If the ServiceInstance or Endpoint is `All`, will serviceInstance and endpoint label pass to the API handler？ Then backend will use `All` as the name to build a not existing Id.
> > 
> > 
> > > 
> > 
> > 
> > > I think `All` in the Grafana UI usually means this field has no query condition, instead of sending an `All` as a query condition.
> > 
> > 
> > The all option in grafana submit all options sperate by `,` and it is not we wanted. So I custom the all value to `blank` and backend filter the blank label value(service, endpoint):
> > image
> 
> Can we customize the all value to `*`? That's much more understandable

If we all think this is better, I feel good with this too.

**Comment by wu-sheng** [comment]
I think for `accurate match` in the query, blank text is fine. `*` is usually used in fuzzy query.

**Comment by weixiang1862** [comment]
> I think for `accurate match` in the query, blank text is fine. `*` is usually used in fuzzy query.
> 
> As @wankai123 and @kezhenxu94 most concern is about readability, you could hard code `*` at the backend to indicate the `all` option. @weixiang1862 You could make the `empty value` works from protocol perspective, as it seems very clear. And supporting `*` to be empty as well to increase the Grafana UI setup experience. We don't have to pick one. :)

Ok, I will  adjust like this.

**Comment by wu-sheng** [comment]
Meanwhile, I think you should be able to send the latest screenshots to the website repo.

**Comment by weixiang1862** [comment]
> Meanwhile, I think you should be able to send the latest screenshots to the website repo.

Ok.

**Review by wu-sheng — DISMISSED** [review]
Screenshots updated. @wankai123 Please check the codes.

**Review by wankai123 — DISMISSED** [review]
LGTM

**Comment by wu-sheng** [comment]
@kezhenxu94 Is the new way good for you?

**Review by kezhenxu94 — DISMISSED** [review]
Just fix some minor typos, LGTM

File: docs/en/changes/changes.md [hunk 7]
```diff
@@ -59,6 +59,7 @@
 * Bump up Armeria to 1.24.3.
 * Apply MQE on APISIX, AWS_EKS, AWS_GATEWAY and AWS_S3 layer UI templates.
 * Fix BooleanMatch and BooleanNotEqualMatch doing Boolean comparison.
+* Support LogQL http query api.
```
Comment (wu-sheng) [inline docs/en/changes/changes.md:62]:
```suggestion
* Support LogQL HTTP query APIs.
```

**Review by wu-sheng — APPROVED** [review]

**Comment by wu-sheng** [comment]
@weixiang1862 I have merged this, you could work on showcase, then we would be able to show this feature on preview demo.

**Comment by weixiang1862** [comment]
> @weixiang1862 I have merged this, you could work on showcase, then we would be able to show this feature on preview demo.

Ok.

**Pull request merged**